Feat/authentication

anmolsahi-mm opened the pull request:
Implemented GitHub OAuth authentication using Firebase

https://user-images.githubusercontent.com/111345322/185842304-731ecff5-2549-493a-b090-a23793490a02.mp4

anmolsahi-mm committed:
Create workflow file to build release apk

anmolsahi-mm committed:
added firebase dependencies

anmolsahi-mm committed:
update java version 11

anmolsahi-mm committed:
Merge branch 'master' of https://github.com/anmolsahi-mm/Github-Clone

anmolsahi-mm committed:
added Firebase Token

anmolsahi-mm committed:
Added Testers group in firebase

anmolsahi-mm committed:
made login screen

anmolsahi-mm committed:
Github Login Working

anmolsahi-mm committed:
Login Flow Completed and added compose Navigation

anmolsahi-mm committed:
Cleaned Up Login Flow

 - Removed the Login Logic from Main Activity
 - Added Common.kt with methods signIn and checkForExistingUser

anmolsahi-mm committed:
Able to get access_token from Github by casting credentials as OAuthCredtenial

anmolsahi-mm committed:
removed extra spaces from MainActivity.kt and deleted unnecessary GithubApplication.kt file

oianmol commented on app/src/main/java/com/example/githubclone/presentation/loginscreen/LoginScreen.kt:
```diff
@@ -0,0 +1,155 @@
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .background(if (isSystemInDarkTheme()) Color(0xFF050505) else Color.White),
```
Should this color come from theme ?

anmolsahi-mm replied:
Yes, I have fixed this and other colors as well in LoginScreen

oianmol commented on app/src/main/java/com/example/githubclone/presentation/loginscreen/LoginScreen.kt:
```diff
@@ -0,0 +1,155 @@
+            ),
+            contentDescription = stringResource(R.string.github_logo),
+            modifier = Modifier
+                .size(82.dp)
```
@anmolsahi-mm https://medium.com/@aratidube12lns/4-point-grid-system-for-more-consistent-interface-design-efea81dea3f3

anmolsahi-mm replied:
Thank you, I have now used 4 point grid system for consistent design

oianmol commented on app/src/main/java/com/example/githubclone/presentation/loginscreen/LoginScreen.kt:
```diff
@@ -0,0 +1,155 @@
+                modifier = Modifier.padding(4.dp),
+                textAlign = TextAlign.Center,
+                style = MaterialTheme.typography.subtitle2,
+                text = buildAnnotatedString {
```
we can get the annotated string from a different function for readability..

anmolsahi-mm replied:
done, moved it to a separate function

shubhamsinghmutualmobile commented on app/src/main/java/com/example/githubclone/MainActivity.kt:
```diff
@@ -6,13 +6,20 @@ import androidx.activity.compose.setContent
 
 class MainActivity : ComponentActivity() {
+
+    private var firebaseUser: FirebaseUser? = null
```
Is this variable being used in this file?

anmolsahi-mm replied:
I have removed this unused variable

shubhamsinghmutualmobile commented on app/src/main/java/com/example/githubclone/navigation/NavGraph.kt:
```diff
@@ -0,0 +1,41 @@
+    NavHost(
+        navController = navController,
+        startDestination =
+        if (firebaseUser != null) Screen.Home.route
```
Could we not simply write something like `Firebase.auth.currentUser != null` instead of creating the `checkForExistingUser` method? Or is it serving some other purpose?

anmolsahi-mm replied:
done, removed this unnecessary method

shubhamsinghmutualmobile commented on app/src/main/java/com/example/githubclone/presentation/common/Common.kt:
```diff
@@ -0,0 +1,83 @@
+import android.util.Log
+import android.widget.Toast
+import com.google.android.gms.tasks.Task
+import com.google.firebase.auth.*
```
Let's not use wildcards (`*`) because they can cause some lint checks to fail (not relevant right now but might be in some other projects). Let's make a habit of using [single imports](https://stackoverflow.com/a/56881421/17621468) instead.

anmolsahi-mm replied:
Thank you, I have changed preferences in Android Studio to not use wildcards

shubhamsinghmutualmobile commented on app/src/main/java/com/example/githubclone/presentation/common/Common.kt:
```diff
@@ -0,0 +1,83 @@
+            val oAuthCredential = authResult.credential as OAuthCredential
+            val accessToken = oAuthCredential.accessToken
+
+            Log.i("Github SignIn", "username: $username, accessToken: $accessToken")
```
We should log only in debug mode.

anmolsahi-mm replied:
Thank you, I have fixed this using Timber

shubhamsinghmutualmobile commented on app/src/main/java/com/example/githubclone/presentation/common/Common.kt:
```diff
@@ -0,0 +1,83 @@
+                activity,
+                "Something Went Wrong: ${exception.message}",
+                Toast.LENGTH_LONG
+            ).show()
```
We can make a generic extension function for displaying toasts (or other common tasks) e.g. 
`fun Context.showToast(msg: String, isLongToast: Boolean = false) {
    Toast.makeText(this, msg, if (isLongToast) Toast.LENGTH_LONG else Toast.LENGTH_SHORT).show()
}`

anmolsahi-mm replied:
done

shubhamsinghmutualmobile commented on app/src/main/java/com/example/githubclone/ui/theme/Shape.kt:
```diff
@@ -5,7 +5,7 @@ import androidx.compose.material.Shapes
 import androidx.compose.ui.unit.dp
 
 val Shapes = Shapes(
```
We can also make use of CompositionLocals for this: https://developer.android.com/jetpack/compose/compositionlocal

shubhamsinghmutualmobile commented on build.gradle:
```diff
@@ -1,6 +1,12 @@
 buildscript {
     ext {
         compose_version = '1.1.0-beta01'
```
Can we migrate to the latest stable version of Compose?

anmolsahi-mm replied:
done, updated to latest compose version

shubhamsinghmutualmobile commented on app/src/main/java/com/example/githubclone/presentation/common/Common.kt:
```diff
@@ -0,0 +1,83 @@
+    val provider: OAuthProvider.Builder = OAuthProvider.newBuilder("github.com")
+    val pendingResultTask: Task<AuthResult>? = firebaseAuth.pendingAuthResult
+
+    if (pendingResultTask != null) {
```
Might not always be the case, but using `nullableVariable?.let { nonNullVariable -> ... } ?: { ... }` is generally safer than using the `if-else` block for checking nulls.

anmolsahi-mm replied:
done, thank you for the info

anmolsahi-mm committed:
fixed review coments
 - made a seperate function to get the Annotated String
 - moved the hard coded colors into Theme
 - used 4 point grid system
 - replaced wildcards with Single imports
 - used Timber for logs in debug mode only
 - migrated to latest stable compose version

anmolsahi-mm committed:
used compose_compiler_version variable
